### ISSUE TITLE
Disentangle the neo datetime datagen

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#5613]
Currently the neo datetime datagen uses the v1 data structs in order to generate v2 data structs. This was a shortest-path solution at the time to get neo up and off the ground, but now it is tech debt, and we should rewrite v2 datagen to read directly from CLDR-JSON data structs.

Also, make sure the word "name" refers to a localized string and "symbol" refers to the placeholder string. https://github.com/unicode-org/icu4x/issues/4328

**Comment by sffc** [comment]
Rust 1.90 makes the `dead_code` lint more aggressive and it has a big impact on this code. We should fix up the code before we can switch the toolchain to 1.90

**Comment by Manishearth** [comment]
I do not think we should block 2.1 on this work. I think we should allow the lint in that module and move on.

Rust updates should not be blocked on large endeavors.

**Comment by sffc** [comment]
I/we rely on the `dead_code` lint heavily. 1.89 to 1.90 makes the lint more aggressive (I would argue _too_ aggressive), and suppressing it, though the quick fix, is _worse_ than the status quo. If we are okay blocking 1.90 but not 2.1 on this issue, we can remove it from the milestone.

**Comment by Manishearth** [comment]
@sffc "We heavily rely on `dead_code`" is not the question IMO. The question is, narrowly: do we rely on `dead_code` _in this file_? This file is a legacy file in need of cleanup: I'd treat this in the same way I'd treat WIP experimental code; in that it is okay to disable lints as long as there is a path to them being fixed.

There is an open, unassigned task to clean this up. It's going to take a fair amount of effort, and when we are doing that, the lint will be valuable to have enabled. _Currently_, none of us are working on this, and the effect of silencing the lint fo this code will be that sometimes when improving nearby code we might leave unused things in this file; this file that is already a mess and in need of targeted cleanup. I think that's fine? I don't think the dead_code lint buys us much for this file.

We may also choose to ask upstream to change the lint. That's fine too. But this is all stuff that can be done separately.


I am reluctant to release a 2.1 when we are not testing with the latest compiler on our CI. That's the type of thing that will come to bite us later if there are other compiler bugs and stuff. We have our nightly tests but that doesn't catch everything.

I'm overall very pleased that ICU4X is able to smoothly keep up to date with the latest Rust stable on our CI, I think that has a lot of good properties and don't enjoy the thought of losing that property.

**Comment by Manishearth** [comment]
In short, my position is:

 - This should not be a 2.1 blocker, especially as an unassigned issue
 - _Ideally_ we are able to release 2.1 with our CI on the latest Rust. I do not think _this_ should block that.

**Comment by Manishearth** [comment]
A corollary is that if we _do_ think the dead_code lint is important enough to this file that we want it enabled, then we should not be landing partial hacks like this at all. I disagree with that: I think landing partial hacks like this is fine, and we treat it like we treat experimental code.

We have landed a partial hack, and the price we pay is that that code will be lower quality for a while. I see that as completely expected and normal, and I don't want the price of that to emanate outwards onto other endeavors, like toolchain management.

**Comment by Manishearth** [comment]
Just ended up seeing if I could sufficiently clean up legacy.rs for the rust update: https://github.com/unicode-org/icu4x/pull/6965

**Comment by Manishearth** [comment]
This is probably mostly done but someone should take a good final look and clean up loose ends.

**Comment by robertbastian** [comment]
We still have some neo.rs modules that'd I'd like to see reorganized semantically.

There's also still an unused legacy.rs module in datagen.

**Comment by Manishearth** [comment]
https://github.com/unicode-org/icu4x/pull/7383 removes the legacy stuff, but it does not attempt to organize neo, yet.
